### ISSUE TITLE
[RFC] Redefine virtual types (`Foo+`) through the notion of polymorphic types

### ISSUE THREAD

**Issue by HertzDevil** [crystal-lang/crystal#10785]
The current behaviour of Crystal's virtual types stems from its very first release when it was still known as [`HierarchyType`](https://github.com/crystal-lang/crystal/blob/8936918be3938f45cda9f0f08bbfff39b8049677/src/compiler/crystal/types.cr#L2505). The logic behind the creation of virtual types is roughly:

* Calling `Crystal::Type#virtual_type` on `Foo` would return either `Foo` or `Foo+` depending on whether it is abstract or has subclasses (and apparently whether it is a struct too).
* `Foo+` is always virtual.
* `Foo` is never virtual.

This distinction is not very helpful, especially the last point. Consider:

```crystal
class Foo; end
class Bar < Foo; end

x = Foo.new || Bar.new
```

Due to Crystal's limitation on struct inheritance, it is impossible that `x` can store a `Foo` but not a `Bar`, for _any_ `Bar < Foo`, whether they are classes, structs, or modules, generic or not. (Contrast with C/C++ where slicing of non-pointer values can occur.) Furthermore, since `Foo` and `Foo+` are not internally equal, this leads to a proliferation of `#devirtualize` calls throughout the compiler in places that shouldn't have to deal with virtual types, e.g. #10302 and #10730.

So I propose to get rid of `Crystal::VirtualType` entirely and define _polymorphic_ and virtual types as follows:

* `Foo` is non-polymorphic if it is a class/struct, has no subclasses, is not abstract, is storable (i.e. can be used as the type of a variable), and is not an uninstantiated generic. A non-polymorphic type has no subtypes, and additionally cannot possibly have one.
* `Foo.class` is non-polymorphic if and only if `Foo` also is.
* A `Tuple` or `NamedTuple` instance is non-polymorphic if and only if all of its member types also are. These two types are more nuanced because they don't have full covariance yet.
* Every other type is polymorphic. In particular, all modules, irreducible unions, and special roots like `Int` would be considered polymorphic. There would be no distinction between `Foo+.class` and `Foo.class+`.
* `Foo` is virtual if at least 2 types among its subtypes and itself are storable. (This moves the concept of virtual types to the codegen phase entirely. It doesn't matter if polymorphic `Tuple`s are virtual since they have their own codegen properties.)

A polymorphic type would then have all the semantic properties of what we currently call virtual types, whereas virtual types would be precisely the types that require type ID dispatches. Virtual implies polymorphic, because a virtual type must have a subtype not equivalent to itself. Their differences are illustrated below:

```crystal
# A is polymorphic because it has a subclass
# A is virtual because A and A1 are both storable
class A; end
class A1 < A; end

# B is polymorphic because it is abstract
# B is virtual because B1 and B2 are both storable
abstract class B; end
class B1 < B; end
class B2 < B; end

# C is polymorphic because it is abstract
# C is non-virtual because A1 is its only storable subtype
abstract class C; end
class C1 < C; end

# D is polymorphic because it is an uninstantiated generic
# D is non-virtual because it has no storable subtypes at all (if no other types include D)
module D(T); end
```

I believe this redefinition of the terms would eliminate a lot of subtleties in the type system, especially in regards to generic inheritance.

**Comment by HertzDevil** [comment]
Another case to consider when always applying the current virtual type semantics to polymorphic types:

```crystal
class A; end
class B < A; end

def foo(x : A)
  # ...
  x.as(B)
  # ...
end

y = A.new
# ...
foo(y)       # Error: can't cast A to B
foo(y.as(A)) # okay, TypeCastError (cast from A to B failed)
```

When one reads `foo`'s declaration, they expect the parameter restriction to behave like `A+` rather than `A`; the compilation error here does not add much value if `y = A.new` and `foo(y)` are placed far apart, perhaps in different files. So I believe the `foo(y)` call should compile, as if `A` has `A+`'s semantics, even if this call always raises during runtime like in the second call to `foo`. (The docs do not mention at all that the `.as(A)` call has the type `A+` instead of `A`; this is just one of the many examples of the compiler generating virtual types beyond union type folding.)

**Comment by asterite** [comment]
I totally agree with you. I wanted to get rid of virtual types for a long time. Well, not of them, but the fact that a type is virtual or not depending on whether they have subclases or are abstract. In the semantic phase we could consider every type as potentially having subtypes. The only thing is, I wonder if compiler performance would be affected. Maybe not.

**Comment by HertzDevil** [comment]
The type inference rules for `as` and `is_a?` both depend on type intersection, and it turns out commutativity fails if the former isn't virtual:

```crystal
class A; end
class B < A; end

a = A.new            # : A
b = B.new            # : B
ua = uninitialized A # : A+
ub = uninitialized B # : B+ = B

typeof(a.is_a?(B) ? a : raise "")   # => NoReturn
typeof(b.is_a?(A) ? b : raise "")   # => B
typeof(ua.is_a?(B) ? ua : raise "") # => B
typeof(ub.is_a?(A) ? ub : raise "") # => B
```

So any discrepancy about `as` also has an `is_a?` counterpart. For example:

```crystal
class C
  def initialize(@x : A)
    y = @x # : A+
    p typeof({
      x.is_a?(B) ? x : raise(""),
      y.is_a?(B) ? y : raise(""),
    })
  end
end

C.new(A.new) # => Tuple(NoReturn, B)
C.new(B.new) # => Tuple(B, B)
```
